### PR TITLE
build: Do not use the configurator for meson

### PR DESCRIPTION
**Pull request by igaw**
meson does the feature detection itself and we don't have to rely on
the configurator. So don't use it.

Signed-off-by: Daniel Wagner <dwagner@suse.de>

### CODE REVIEW

**Comment by igaw** [comment]
Also relevant: #124

**Comment by hreinecke** [comment]
Can't we just leave the 'ccan' header files intact (with config.h etc), and just remove ccan/meson.build?

**Comment by igaw** [comment]
This is possible but then we have to rename libnvme-config.h  back to config.h and and move it out of the default include directory to avoid include collisions. Hmm, IIRC we should not include the the '.' anyway. I agree it would be better not to touch the ccan files. Let me play with this.

**Comment by igaw** [comment]
This seems approach seems to work with nvme-cli.

**Comment by igaw** [comment]
Here the some output for a build with nvme-cli
- libnvme
> cc -Isubprojects/libnvme/src/libnvme.so.1.0.0.p -Isubprojects/libnvme/src -I../subprojects/libnvme/src -Isubprojects/libnvme/ccan -I../subprojects/libnvme/ccan -I/usr/include/uuid -I/usr/include/json-c -fdiagnostics-color=always -D_FILE_OFFSET_BITS=64 -Wall -Winvalid-pch -std=gnu99 -O3 -g -fomit-frame-pointer -D_GNU_SOURCE -include config.h -fPIC -MD -MQ subprojects/libnvme/src/libnvme.so.1.0.0.p/nvme_json.c.o -MF subprojects/libnvme/src/libnvme.so.1.0.0.p/nvme_json.c.o.d -o subprojects/libnvme/src/libnvme.so.1.0.0.p/nvme_json.c.o -c ../subprojects/libnvme/src/nvme/json.c

This call doesn't include the '-I.' path which means when compiling the subproject the config file from '-I../subprojects/libnvme/src' is used. Looks about right

- nvme-cli
> cc -Invme.p -I. -I.. -Iccan -I../ccan -Isubprojects/libnvme/ccan -I../subprojects/libnvme/ccan -Isubprojects/libnvme/src -I../subprojects/libnvme/src -I/usr/include/uuid -I/usr/include/json-c -fdiagnostics-color=always -D_FILE_OFFSET_BITS=64 -Wall -Winvalid-pch -std=gnu99 -O3 -g -fomit-frame-pointer -D_GNU_SOURCE -include config.h -DOPENSSL_LOAD_CONF -MD -MQ nvme.p/nvme-print.c.o -MF nvme.p/nvme-print.c.o.d -o nvme.p/nvme-print.c.o -c ../nvme-print.c

The first include '-I.' will include the config.h generated by nvme-cli. If I am not completely misstaken this will used before the one under '-Isubprojects/libnvme/src'. This should work for the nvme-cli

**Comment by igaw** [comment]
Build passes with 'meson .build -Db_sanitize=address' too.

**Comment by igaw** [comment]
There is some problems with 'meson .build -Db_sanitize=address && ninja -C .build test':

```
6/7 [Python] import libnvme            FAIL            0.03s   exit status 1
>>> PYTHONPATH=/home/wagi/work/libnvme/.build/libnvme/.. MALLOC_PERTURB_=0 /usr/bin/python3 -c 'from libnvme import nvme'
――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――― ✀  ――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――
stderr:
==29060==ASan runtime does not come first in initial library list; you should either link runtime to your application or manually preload it with LD_PRELOAD.

```
This is not new failure. I think it's due the current way we setup the Python tests.

**Comment by martin-belanger** [comment]
IIRC, when using the sanitize option `ld` has to link against `libasan.so`. And this library has to come first in the library list. This is what the message is about. Let me do some Google'ing to see how to fix this.

**Comment by martin-belanger** [comment]
Found the solution. When using the sanitizer, the very first library loaded must be `libasan.so`. This can be achieved using the `LD_PRELOAD` environment variable. In other words, the following works:
```
meson .build -Db_sanitize=address && LD_PRELOAD=/lib64/libasan.so.6 ninja -C .build test 
```

And this is the output of the test:
```
ninja: Entering directory `.build'
[0/1] Running all tests.
1/7 main                               OK              0.03s
2/7 main                               OK              0.06s
3/7 cpp                                OK              0.07s
4/7 register                           OK              0.07s
5/7 zns                                OK              0.05s
6/7 [Python] import libnvme            FAIL            0.18s   exit status 1
>>> MALLOC_PERTURB_=0 PYTHONPATH=/mnt/conundrum/home/mbelanger/work/nvme/libnvme/.build/libnvme/.. /usr/bin/python3 -c 'from libnvme import nvme'
――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――― ✀  ――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――
stderr:

=================================================================
==6653==ERROR: LeakSanitizer: detected memory leaks

Direct leak of 167102 byte(s) in 95 object(s) allocated from:
    #0 0x7fe28983995f in __interceptor_malloc (/lib64/libasan.so.6+0xae95f)
    #1 0x7fe28952bdae in PyObject_Malloc (/lib64/libpython3.10.so.1.0+0xf7dae)

Indirect leak of 13469 byte(s) in 13 object(s) allocated from:
    #0 0x7fe28983995f in __interceptor_malloc (/lib64/libasan.so.6+0xae95f)
    #1 0x7fe28952bdae in PyObject_Malloc (/lib64/libpython3.10.so.1.0+0xf7dae)

SUMMARY: AddressSanitizer: 180571 byte(s) leaked in 108 allocation(s).
―――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――

7/7 [Python] create ctrl object        FAIL            0.18s   exit status 1
>>> MALLOC_PERTURB_=0 PYTHONPATH=/mnt/conundrum/home/mbelanger/work/nvme/libnvme/.build/libnvme/.. /usr/bin/python3 /mnt/conundrum/home/mbelanger/work/nvme/libnvme/.build/../libnvme/tests/create-ctrl-obj.py
――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――― ✀  ――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――
stderr:

=================================================================
==6655==ERROR: LeakSanitizer: detected memory leaks

Direct leak of 168174 byte(s) in 95 object(s) allocated from:
    #0 0x7f03f3fe195f in __interceptor_malloc (/lib64/libasan.so.6+0xae95f)
    #1 0x7f03f3cd3dae in PyObject_Malloc (/lib64/libpython3.10.so.1.0+0xf7dae)

Direct leak of 1048 byte(s) in 2 object(s) allocated from:
    #0 0x7f03f3fe195f in __interceptor_malloc (/lib64/libasan.so.6+0xae95f)
    #1 0x7f03f3d58595 in _PyObject_Malloc.part.0 (/lib64/libpython3.10.so.1.0+0x17c595)

Indirect leak of 13469 byte(s) in 13 object(s) allocated from:
    #0 0x7f03f3fe195f in __interceptor_malloc (/lib64/libasan.so.6+0xae95f)
    #1 0x7f03f3cd3dae in PyObject_Malloc (/lib64/libpython3.10.so.1.0+0xf7dae)

SUMMARY: AddressSanitizer: 182691 byte(s) leaked in 110 allocation(s).
―――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――


Summary of Failures:

6/7 [Python] import libnvme     FAIL            0.18s   exit status 1
7/7 [Python] create ctrl object FAIL            0.18s   exit status 1


Ok:                 5   
Expected Fail:      0   
Fail:               2   
Unexpected Pass:    0   
Skipped:            0   
Timeout:            0   

Full log written to /mnt/conundrum/home/mbelanger/work/nvme/libnvme/.build/meson-logs/testlog.txt

=================================================================
==6652==ERROR: LeakSanitizer: detected memory leaks

Direct leak of 106724 byte(s) in 90 object(s) allocated from:
    #0 0x7f7c9e9a295f in __interceptor_malloc (/lib64/libasan.so.6+0xae95f)
    #1 0x7f7c9e694dae in PyObject_Malloc (/lib64/libpython3.10.so.1.0+0xf7dae)

Indirect leak of 11248 byte(s) in 10 object(s) allocated from:
    #0 0x7f7c9e9a295f in __interceptor_malloc (/lib64/libasan.so.6+0xae95f)
    #1 0x7f7c9e694dae in PyObject_Malloc (/lib64/libpython3.10.so.1.0+0xf7dae)

SUMMARY: AddressSanitizer: 117972 byte(s) leaked in 100 allocation(s).
FAILED: meson-test 
/usr/bin/meson test --no-rebuild --print-errorlogs
ninja: build stopped: subcommand failed.

=================================================================
==6651==ERROR: LeakSanitizer: detected memory leaks

Direct leak of 296 byte(s) in 1 object(s) allocated from:
    #0 0x7f4dc18f40c7 in operator new(unsigned long) (/lib64/libasan.so.6+0xb00c7)
    #1 0x55ed22c9b065 in Builder::Builder(State*, BuildConfig const&, BuildLog*, DepsLog*, DiskInterface*) (/usr/bin/ninja+0x9065)
    #2 0x55ed22c9781e in main (/usr/bin/ninja+0x581e)

Indirect leak of 512 byte(s) in 1 object(s) allocated from:
    #0 0x7f4dc18f40c7 in operator new(unsigned long) (/lib64/libasan.so.6+0xb00c7)
    #1 0x55ed22c9b324 in Builder::Builder(State*, BuildConfig const&, BuildLog*, DepsLog*, DiskInterface*) (/usr/bin/ninja+0x9324)

Indirect leak of 64 byte(s) in 1 object(s) allocated from:
    #0 0x7f4dc18f40c7 in operator new(unsigned long) (/lib64/libasan.so.6+0xb00c7)
    #1 0x55ed22c9b2fc in Builder::Builder(State*, BuildConfig const&, BuildLog*, DepsLog*, DiskInterface*) (/usr/bin/ninja+0x92fc)

SUMMARY: AddressSanitizer: 872 byte(s) leaked in 3 allocation(s).

```

File: src/meson.build [hunk 1]
```diff
@@ -5,6 +5,11 @@
 #
 # Authors: Martin Belanger <Martin.Belanger@dell.com>
 #
+configure_file(
+    output: 'config.h',
+    configuration: conf
+)
+
 sources += [
```
Comment (martin-belanger) [inline src/meson.build:13]: Hi @igaw 

We could directly add the ccan source files here and completely eliminate the need for `ccan/meson.build`.

For example:
```
sources = [
    '../ccan/ccan/list/list.c',
    '../ccan/ccan/str/debug.c',
    '../ccan/ccan/str/str.c',
    'nvme/cleanup.c',
    'nvme/fabrics.c',
    'nvme/filters.c',
    'nvme/ioctl.c',
    'nvme/linux.c',
    'nvme/log.c',
    'nvme/tree.c',
    'nvme/util.c',
]
```
I can submit a pull request if you'd like.

Comment (igaw) [reply src/meson.build:13]: I don't have any strong feelings on this. If the general mode is that we need to get rid of the ccan/meson.build, please go ahead.

**Comment by igaw** [comment]
About the LD_LIBRARY preload trick. It think this could be added to the REAMDE.

File: src/meson.build (hunk 1 above)
Comment (martin-belanger) [reply src/meson.build:13]: This is weird. The 3 ccan files that are added to `sources` are actually not needed at all. I cannot find any libnvme code that uses any of the functions defined in these 3 files. I just tried building without these files and I ran all the tests successfully. I will submit a pull request that cleans all of that up.

**Pull request merged**